File: src/test/java/com/github/kayjamlang/tests/expressions/WhileExpressionTest.java
package com.github.kayjamlang.tests.expressions;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.ValueExpression;
import com.github.kayjamlang.core.expressions.loops.WhileExpression;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class WhileExpressionTest {

    private static KayJamParser parser;

    @BeforeClass
    public static void prepare(){
        parser = new KayJamParser(new KayJamLexer("while(false) true"));
    }

    @Test
    public void test() throws Exception {
        Expression expression = parser.readExpression();

        assertNotNull(expression);
        assertSame(WhileExpression.class, expression.getClass());

        WhileExpression whileExpression = (WhileExpression) expression;
        assertSame(ValueExpression.class, whileExpression.condition.getClass());
        assertSame(ValueExpression.class, whileExpression.expression.getClass());

        ValueExpression condition = (ValueExpression) whileExpression.condition;
        assertEquals(false, condition.value);

        ValueExpression cValue = (ValueExpression) whileExpression.expression;
        assertEquals(true, cValue.value);
    }
}

File: src/test/scala/com/github/kayjamlang/tests/expressions/WhileExpressionTest.scala
package com.github.kayjamlang.tests.expressions

import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.ValueExpression
import com.github.kayjamlang.core.expressions.loops.WhileExpression
import org.junit.Assert._
import org.junit.{BeforeClass, Test}

object WhileExpressionTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer(
      """
        |while (false)
        | true
        |""".stripMargin))
  }
}

class WhileExpressionTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = WhileExpressionTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[WhileExpression], expression getClass)
    val whileExpression = expression.asInstanceOf[WhileExpression]
    assertSame(classOf[ValueExpression], whileExpression.condition getClass)
    assertSame(classOf[ValueExpression], whileExpression.expression getClass)
    val condition = whileExpression.condition.asInstanceOf[ValueExpression]
    assertEquals(false, condition.value)
    val cValue = whileExpression.expression.asInstanceOf[ValueExpression]
    assertEquals(true, cValue.value)
  }
}